cmd/otel-allocator/collector/collector.go

```diff
@@ -43,110 +43,102 @@ var (
-			return nil
+	notifyFunc := func(_ interface{}) {
+		select {
+		case notify <- struct{}{}:
+		default:
 		}
 	}
-}
-
-func (k *Client) restartWatch(ctx context.Context, opts metav1.ListOptions, collectorMap map[string]*allocation.Collector, fn func(collectors map[string]*allocation.Collector)) bool {
-	// add timeout to the context before calling Watch
-	ctx, cancel := context.WithTimeout(ctx, watcherTimeout)
-	defer cancel()
-	watcher, err := k.k8sClient.CoreV1().Pods(ns).Watch(ctx, opts)
+	_, err = informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: notifyFunc,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			notifyFunc(newObj)
+		},
+		DeleteFunc: notifyFunc,
+	})
 	if err != nil {
-		k.log.Error(err, "unable to create collector pod watcher")
-		return false
-	}
-	k.log.Info("Successfully started a collector pod watcher")
-	if msg := runWatch(ctx, k, watcher.ResultChan(), collectorMap, fn); msg != "" {
-		k.log.Info("Collector pod watch event stopped " + msg)
-		return false
+		return err
 	}
 
-	return true
+	informer.Run(k.close)
+	return nil
 }
 
-func runWatch(ctx context.Context, k *Client, c <-chan watch.Event, collectorMap map[string]*allocation.Collector, fn func(collectors map[string]*allocation.Collector)) string {
+// rateLimitedCollectorHandler runs fn on collectors present in the store whenever it gets a notification on the notify channel,
+// but not more frequently than once per k.eventPeriod.
+func (k *CollectorWatcher) rateLimitedCollectorHandler(notify chan struct{}, store cache.Store, fn func(collectors map[string]*allocation.Collector)) {
+	ticker := time.NewTicker(k.minUpdateInterval)
+	defer ticker.Stop()
+
 	for {
-		collectorsDiscovered.Set(float64(len(collectorMap)))
 		select {
 		case <-k.close:
-			return "kubernetes client closed"
-		case <-ctx.Done():
-			return ""
-		case event, ok := <-c:
-			if !ok {
-				k.log.Info("No event found. Restarting watch routine")
-				return ""
-			}
-
-			pod, ok := event.Object.(*v1.Pod)
-			if !ok {
-				k.log.Info("No pod found in event Object. Restarting watch routine")
-				return ""
-			}
-
-			if pod.Spec.NodeName == "" {
-				k.log.Info("Node name is missing from the spec. Restarting watch routine")
-				return ""
+			return
+		case <-ticker.C: // throttle events to avoid excessive updates
+			select {
+			case <-notify:
+				k.runOnCollectors(store, fn)
+			default:
 			}
+		}
+	}
+}
 
-			switch event.Type { //nolint:exhaustive
-			case watch.Added:
-				collectorMap[pod.Name] = allocation.NewCollector(pod.Name, pod.Spec.NodeName)
-			case watch.Deleted:
-				delete(collectorMap, pod.Name)
-			}
-			fn(collectorMap)
+// runOnCollectors runs the provided function on the set of collectors from the Store.
+func (k *CollectorWatcher) runOnCollectors(store cache.Store, fn func(collectors map[string]*allocation.Collector)) {
+	collectorMap := map[string]*allocation.Collector{}
+	objects := store.List()
```

Comment:
It will return all existing Collector Pods, not just active ones. Below, we also filter it down to scheduled ones by requiring that `NodeName` be present. We do not check for readiness here - I'd like to do that in a follow up. This behaviour is exactly what we currently have.